PhysicsTools/PatAlgos/plugins/PATIsolatedTrackProducer.cc

```diff
@@ -1,92 +1,377 @@
+    edm::Handle<reco::DeDxHitInfoAss> gt2dedxHitInfo;
+    iEvent.getByToken(gt2dedxHitInfo_, gt2dedxHitInfo);
+
+    edm::ESHandle<HcalChannelQuality> hcalQ_h;
+    iSetup.get<HcalChannelQualityRcd>().get("withTopo", hcalQ_h);
+    const HcalChannelQuality *hcalQ = hcalQ_h.product();
     
-    for(pat::PackedCandidateCollection::const_iterator pf_it = pc->begin(); pf_it != pc->end(); pf_it++){
+    edm::ESHandle<EcalChannelStatus> ecalS_h;
+    iSetup.get<EcalChannelStatusRcd>().get(ecalS_h);
+    const EcalChannelStatus *ecalS = ecalS_h.product();
+
+    auto outPtrP = std::make_unique<std::vector<pat::IsolatedTrack>>();
+
+    //add general tracks
+    for(unsigned int igt=0; igt<generalTracks->size(); igt++){
+        const reco::Track &gentk = (*gt_h)[igt];
+        reco::TrackRef tkref = reco::TrackRef(gt_h, igt);
+        pat::PackedCandidateRef pcref = (*gt2pc)[tkref];
+        pat::PackedCandidateRef ltref = (*gt2lt)[tkref];
+
+        // Determine if this general track is associated with anything in packedPFCandidates or lostTracks
+        // Sometimes, a track gets associated w/ a neutral pfCand.
+        // In this case, ignore the pfCand and take from lostTracks
+        bool isInPackedCands = (pcref.isNonnull() && pcref.id()==pc_h.id() && pcref.get()->charge()!=0);
+        bool isInLostTracks  = (ltref.isNonnull() && ltref.id()==lt_h.id());
+
+        LorentzVector p4;
+        pat::PackedCandidateRef refToCand;
+        int pdgId, charge, fromPV;
+        float dz, dxy, dzError, dxyError;
+        int pfCandInd; //to avoid counting packedPFCands in their own isolation
+
+        // get the four-momentum and charge
+        if(isInPackedCands){
+            p4 = pcref.get()->p4();
+            charge = pcref.get()->charge();
+            pfCandInd = pcref.key();
+        }else if(isInLostTracks){
+            p4 = ltref.get()->p4();
+            charge = ltref.get()->charge();
+            pfCandInd = -1;
+        }else{
+            double m = 0.13957018; //assume pion mass
+            double E = sqrt(m*m + gentk.p()*gentk.p());
+            p4.SetPxPyPzE(gentk.px(), gentk.py(), gentk.pz(), E);
+            charge = gentk.charge();
+            pfCandInd = -1;
+        }
+
+        if(p4.pt() < pT_cut_)
+            continue;
+        if(charge == 0)
+            continue;
+
+        // get the isolation of the track
+        pat::PFIsolation isolationDR03;
+        pat::PFIsolation miniIso;
+        getIsolation(p4, pc, pfCandInd, isolationDR03, miniIso);
+        
+        // isolation cut
+        if( p4.pt() < pT_cut_noIso_ && 
+            !(isolationDR03.chargedHadronIso() < absIso_cut_ ||
+              isolationDR03.chargedHadronIso()/p4.pt() < relIso_cut_ ||
+              miniIso.chargedHadronIso()/p4.pt() < miniRelIso_cut_))
+            continue;
+
+        // get the rest after the pt/iso cuts. Saves some runtime
+        if(isInPackedCands){
+            pdgId = pcref.get()->pdgId();
```

Comment:
```auto const& pc = *(pcref.get());``` could save some function copy-paste and not rely on the compiler to optimize out the repeated calls